src/Runner.Listener/Runner.cs

```diff
@@ -474,10 +475,37 @@ private async Task<int> RunAsync(RunnerSettings settings, bool runOnce = false)
                                     var credMgr = HostContext.GetService<ICredentialManager>();
                                     var creds = credMgr.LoadCredentials();
 
-                                    // todo: add retries https://github.com/github/actions-broker/issues/49
                                     var runServer = HostContext.CreateService<IRunServer>();
                                     await runServer.ConnectAsync(new Uri(settings.ServerUrl), creds);
-                                    var jobMessage = await runServer.GetJobMessageAsync(messageRef.RunnerRequestId);
+                                    Pipelines.AgentJobRequestMessage jobMessage = null;
+
+                                    var remainingRetryLimitTime = TimeSpan.FromMinutes(5);
+                                    while (true)
+                                    {
+                                        try
+                                        {
+                                            jobMessage = await runServer.GetJobMessageAsync(messageRef.RunnerRequestId);
```

Comment:
Why do we need a retry at this layer? 
- the VSSHttpClient already has retry built-in.
- I don't think we want to retry on all exceptions
- If we really want to retry, I would suggest retrying within the `runServer.GetJobMessageAsync` to keep the `Runner.cs` simple.

Reply:
@TingluoHuang that is interesting. Keep in mind VSSF is temporary. Eventually will switch over to Run svc.

We could avoid retrying some HTTP codes (400/401/403) but due to eventual consistency edge cases, i think we want to retry 404.

With all ☝️ in mind, do you think we need to add any extra retry logic at this time? I suppose we could simply wait until we switch to Run svc.

Reply:
@TingluoHuang @ericsciple I rewrite a bit.

BTW, I had some other draft solution: https://github.com/actions/runner/pull/1938